clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -672,6 +672,96 @@ check_function_markers()
+            // Return from function 2.
+            gen_instr_type(TRACE_TYPE_INSTR_RETURN, TID, FUNC2_BASE + 8, RETURN_SZ),
+
+            gen_instr(TID, FUNC1_BASE + CALL_SZ, 8),
+            // Jump back to the beginning of function 1.
+            gen_instr_type(TRACE_TYPE_INSTR_TAKEN_JUMP, TID, FUNC1_BASE + CALL_SZ + 8),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_ID, 1),
+            // The return address should be the same as the return address of
+            // the original call to function 1.
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_RETADDR, BASE_ADDR + CALL_SZ),
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Correctly handle kernel transfer, sigreturn, nested function calls
+    // including tailcalls.
+    {
+        constexpr addr_t BASE_ADDR = 100;
+        constexpr addr_t FUNC1_BASE = 200;
+        constexpr addr_t FUNC2_BASE = 300;
+        constexpr addr_t SIG_HANDLER_BASE = 400;
+        constexpr addr_t SYSCALL_BASE = 500;
+        constexpr size_t RETURN_SZ = 3;
+        constexpr size_t SYSCALL_SZ = 2;
+
+        std::vector<memref_t> memrefs = {
+            gen_instr(TID, 1, BASE_ADDR),
+            // kernel xfer.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, BASE_ADDR + 1),
+            gen_marker(TID, TRACE_MARKER_TYPE_TIMESTAMP, 6),
+            gen_marker(TID, TRACE_MARKER_TYPE_CPU_ID, 3),
+            // Call function 1.
+            gen_instr_type(TRACE_TYPE_INSTR_DIRECT_CALL, TID, SIG_HANDLER_BASE, CALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_ID, 1),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_RETADDR, SIG_HANDLER_BASE + CALL_SZ),
+            // Call function 2.
+            gen_instr_type(TRACE_TYPE_INSTR_DIRECT_CALL, TID, FUNC1_BASE, CALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_ID, 2),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_RETADDR, FUNC1_BASE + CALL_SZ),
+            // Return from function 2.
+            gen_instr_type(TRACE_TYPE_INSTR_RETURN, TID, FUNC2_BASE, RETURN_SZ),
+            // Return from function 1.
+            gen_instr_type(TRACE_TYPE_INSTR_RETURN, TID, FUNC1_BASE + CALL_SZ, RETURN_SZ),
+            // Return from the signal handler.
+            gen_instr_type(TRACE_TYPE_INSTR_RETURN, TID, SIG_HANDLER_BASE + CALL_SZ,
+                           RETURN_SZ),
+            // sigreturn.
+            gen_instr(TID, SYSCALL_SZ, SYSCALL_BASE),
+            gen_marker(TID, TRACE_MARKER_TYPE_SYSCALL, 15),
+            gen_marker(TID, TRACE_MARKER_TYPE_TIMESTAMP, 16),
+            gen_marker(TID, TRACE_MARKER_TYPE_CPU_ID, 3),
+            // syscall xfer.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_XFER, SYSCALL_BASE + SYSCALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_TIMESTAMP, 17),
+            gen_marker(TID, TRACE_MARKER_TYPE_CPU_ID, 3),
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
```

Comment:
It may be worth testing a nested signal and a back-to-back signal as well.

Reply:
I've added a test for nested signal, and another one for back to back signal.